### PR TITLE
[10.0][IMP] sale_commission_formula: Compute commissions using unified method

### PR DESCRIPTION
**Pull request by SimoRubi**
The unified method being `_get_commission_amount`.
Before this PR, for each line all the lines in `self` were recomputed too

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Why this is not done in `sale.commission.line.mixin` object and remove the other 2 overrides (in invoice and in sales order)?

**Comment by SimoRubi** [comment]
> Why this is not done in `sale.commission.line.mixin` object and remove the other 2 overrides (in invoice and in sales order)?

Thanks @pedrobaeza for the review, I didn't notice that there was the same code in `sale.order` too, done

File: sale_commission_formula/models/sale_commission_mixin.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    def _get_commission_amount(
+            self, commission, subtotal, commission_free, product, quantity):
+        self.ensure_one()
+        if commission.commission_type == 'formula' and \
+                not commission_free and commission:
+            formula = commission.formula
+            results = self._get_formula_input_dict()
+            safe_eval(formula, results, mode="exec", nocopy=True)
+            return float(results['result'])
+        else:
```
Comment (pedrobaeza) [inline sale_commission_formula/models/sale_commission_mixin.py]: No need of `else`

File: sale_commission_formula/models/sale_commission_mixin.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from odoo import api, models
+from odoo.tools.safe_eval import safe_eval
+
+
+class SaleCommissionMixin(models.AbstractModel):
+    _inherit = 'sale.commission.mixin'
+
+    @api.model
+    def _get_formula_input_dict(self):
+        return {'line': self.object_id,
```
Comment (pedrobaeza) [inline sale_commission_formula/models/sale_commission_mixin.py]: Why do you overwrite this?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: sale_commission_formula/models/sale_commission_mixin.py (hunk 2 above)
Comment (SimoRubi) [reply sale_commission_formula/models/sale_commission_mixin.py]: @pedrobaeza I was right now testing this locally, I have overridden `sale.commission.mixin` by mistake, the correct model to override is `sale.commission.line.mixin`, fixed in https://github.com/OCA/commission/compare/6e24d9bad74ee98e56cba5be6ee4a95f43865341..6e998cdfb1e42ba552e154bb4e92b6033376a808

File: sale_commission_formula/models/sale_commission_mixin.py (hunk 1 above)
Comment (SimoRubi) [reply sale_commission_formula/models/sale_commission_mixin.py]: @pedrobaeza I am calling super to get the correct `amount` if the `commission_type` is not `formula`.

File: sale_commission_formula/models/sale_commission_line_mixin.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    def _get_commission_amount(
+            self, commission, subtotal, commission_free, product, quantity):
+        self.ensure_one()
+        if commission.commission_type == 'formula' and \
+                not commission_free and commission:
+            formula = commission.formula
+            results = self._get_formula_input_dict()
+            safe_eval(formula, results, mode="exec", nocopy=True)
+            return float(results['result'])
+        else:
```
Comment (pedrobaeza) [inline sale_commission_formula/models/sale_commission_line_mixin.py]: What I mean is that you can put `super` call without the else.

File: sale_commission_formula/models/sale_commission_line_mixin.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+from odoo import api, models
+from odoo.tools.safe_eval import safe_eval
+
+
+class SaleCommissionLineMixin(models.AbstractModel):
+    _inherit = 'sale.commission.line.mixin'
+
+    @api.model
+    def _get_formula_input_dict(self):
```
Comment (pedrobaeza) [inline sale_commission_formula/models/sale_commission_line_mixin.py:13]: Again, this shouldn't be needed

File: sale_commission_formula/models/sale_commission_line_mixin.py (hunk 3 above)
Comment (SimoRubi) [reply sale_commission_formula/models/sale_commission_line_mixin.py]: Oh, ok I didn't get that, done

File: sale_commission_formula/models/sale_commission_line_mixin.py (hunk 4 above)
Comment (SimoRubi) [reply sale_commission_formula/models/sale_commission_line_mixin.py:13]: I'm sorry but I'm not getting what you mean here.
The method is created to be overridden from whoever might be willing to add values for the formula to use. 
I can't find any other occurrence of `_get_formula_input_dict` so it is not overriding anything

Comment (SimoRubi) [reply sale_commission_formula/models/sale_commission_line_mixin.py:13]: On a side note, I just edited that to use the correct decorator `api.multi` with `ensure_one`

**Comment by SimoRubi** [comment]
Travis's automated tests have an error not depending on this, to be fixed in https://github.com/OCA/commission/pull/208

File: sale_commission_formula/models/sale_commission_line_mixin.py (hunk 4 above)
Comment (pedrobaeza) [reply sale_commission_formula/models/sale_commission_line_mixin.py:13]: OK, got it. I thought this method was declared already on super. My fault.

**Review by pedrobaeza — APPROVED** [review]

**Comment by SimoRubi** [comment]
Rebased after #208

**Review by eLBati — APPROVED** [review]

**Pull request merged**